server/src/main/java/org/elasticsearch/node/NodeConstruction.java

```diff
@@ -1087,11 +1089,27 @@ record PluginServiceInstances(
             );
         }
 
+        // Register noop versions of inference services if Inference plugin is not available
+        if (isPluginComponentDefined(pluginComponents, InferenceServiceRegistry.class) == false) {
```

Comment:
I'm not happy about this - feels quite hacky. We don't have a way for Guice to accept a binding to not be present and need to provide a default binding in that case.

The alternative I can think of is something similar to [telemetry plugin](https://github.com/elastic/elasticsearch/blob/6539c2a8efc84bd398b5ecaa23292d9eb140b144/server/src/main/java/org/elasticsearch/node/NodeConstruction.java#L457) - create a specific interface (`InferenceRegistryPlugin`) for retrieving the  `ModelRegistry` and  `InferenceServiceRegistry`, and let InferencePlugin implement it.

Thoughts?

Reply:
If the `InferencePlugin` is present then `ModelRegistry` and `InferenceServiceRegistry` will come from `InferencePlugin::createComponents` otherwise the NOOP implementations are created here. I think the pattern used in the telemetry plugin makes that connection explicit and is less confusing if you don't know the codebase. For that reason I prefer your alternative suggestion of filter for an `InferencePlugin` (there should only be one so log an error if > 1) and ask the `InferencePlugin` for the `ModelRegistry` and `InferenceServiceRegistry` implementations.

Reply:
I just implemented that, @davidkyle . Can you please another look at this? 🙏